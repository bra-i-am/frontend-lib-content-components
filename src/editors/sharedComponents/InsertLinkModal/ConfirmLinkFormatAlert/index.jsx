import { useState } from 'react';
import PropTypes from 'prop-types';
import { useIntl } from '@edx/frontend-platform/i18n';
import {
  Form,
  AlertModal,
  ActionRow,
  Button,
} from '@openedx/paragon';
import { Warning } from '@openedx/paragon/icons';

import messages from './messages';

const ConfirmLinkFormatAlert = ({
  url,
  onCloseAlert,
  editorRef,
}) => {
  const intl = useIntl();
  const [protocol, setProtocol] = useState('');
  const handleChangeProtocol = e => setProtocol(e.target.value);

  const handleCancel = () => onCloseAlert();

  /* istanbul ignore next */
  const handleConfirm = () => {
    const editor = editorRef.current;
    if (editor) {
      const selectedText = editor.selection.getContent({ format: 'text' });
      const urlPath = `${protocol}${url}`;

      if (selectedText.trim() !== '') {
        const linkHtml = `<a href="${urlPath}" data-mce-href="${urlPath}" target="_blank">${selectedText}</a>`;
        editor.selection.setContent(linkHtml);
      }
    }
    onCloseAlert();
  };

  return (
    <AlertModal
      title={intl.formatMessage(messages.confirmLinkFormatAlertTitle)}
      isOpen
      onClose={onCloseAlert}
      variant="warning"
      icon={Warning}
      footerNode={(
        <ActionRow>
          <Button
            variant="tertiary"
            onClick={handleCancel}
            data-testid="cancel-button"
          >
            {intl.formatMessage(messages.confirmLinkFormatAlertCancel)}
          </Button>
          <Button
            variant="primary"
            disabled={!protocol}
            data-testid="confirm-button"
            onClick={handleConfirm}
          >
            {intl.formatMessage(messages.confirmLinkFormatAlertConfirm)}
          </Button>
        </ActionRow>
      )}
    >
      <Form.Group>
        <Form.Label className="my-4">
          {intl.formatMessage(messages.confirmLinkFormatAlertDescription)}
        </Form.Label>
        <Form.RadioSet
          name="colors"
          onChange={handleChangeProtocol}
          value={protocol}
        >
          <Form.Radio value="http://" data-testid="radio-http">
            {intl.formatMessage(messages.confirmLinkFormatAlertHttpProtocol)}
          </Form.Radio>
          <Form.Radio value="https://" data-testid="radio-https">
            {intl.formatMessage(messages.confirmLinkFormatAlertHttpsProtocol)}
          </Form.Radio>
        </Form.RadioSet>
      </Form.Group>
    </AlertModal>
  );
};

ConfirmLinkFormatAlert.propTypes = {
  url: PropTypes.string.isRequired,
  onCloseAlert: PropTypes.func.isRequired,
  editorRef: PropTypes.shape({
    current: PropTypes.shape({
      selection: PropTypes.shape({
        getContent: PropTypes.func,
        setContent: PropTypes.func,
      }),
    }),
  }).isRequired,
};

export default ConfirmLinkFormatAlert;
