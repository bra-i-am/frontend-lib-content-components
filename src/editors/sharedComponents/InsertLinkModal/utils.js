import cloneDeep from 'lodash.clonedeep';

export const blockTypes = {
  section: 'chapter',
  subsection: 'sequential',
  unit: 'vertical',
};

/**
 * Recursively adds path, parent ID, and root status to blocks in a nested structure.
 *
 * @param {Object} block - The current block in the recursion.
 * @param {string} [parentPath=""] - The path of the parent block.
 * @param {Object} blocks - The collection of blocks.
 * @param {string} blockRoot - The key of the root block.
 * @param {string|null} [parentId=null] - The ID of the parent block.
 */
export const addPathToBlocks = (block, blocks, blockRoot, parentId = null, parentPath = '') => {
  const path = parentPath ? `${parentPath} / ${block.displayName}` : block.displayName;
  block.path = path;
  block.parentId = parentId;

  if (block.children && block.children.length > 0) {
    block.children.forEach(childKey => {
      const childBlock = blocks[childKey];
      addPathToBlocks(childBlock, blocks, blockRoot, block.id, path);
    });
  }
};

/**
 * Formats the blocks by adding path information to each block.
 *
 * @param {Object} blocks - The blocks to be formatted.
 * @param {string} blockRoot - The key of the root block.
 * @returns {Object} - The formatted blocks with added path information.
 */
export const formatBlocks = (blocks, blockRoot) => {
  const copyBlocks = cloneDeep(blocks);
  Object.keys(copyBlocks).forEach(key => {
    const block = copyBlocks[key];
    const rootBlock = copyBlocks[blockRoot];
    const parentPath = block.type === blockTypes.section ? rootBlock.displayName : '';

    addPathToBlocks(block, copyBlocks, blockRoot, null, parentPath);
  });

  return copyBlocks;
};

/**
 * Retrieves a list of sections from the provided blocks object.
 *
 * @param {Object} blocks - The blocks object containing various block types.
 * @returns {Array} An array of section (type: chapter) blocks extracted from the blocks object.
 */
export const getSectionsList = (blocks = {}) => {
  const blocksList = Object.keys(blocks);
  return blocksList.reduce((previousBlocks, blockKey) => {
    const block = cloneDeep(blocks[blockKey]);
    if (block.type === blockTypes.section) {
      return [...previousBlocks, block];
    }

    return previousBlocks;
  }, []);
};

/**
 * Retrieves an array of child blocks based on the children list of a selected block.
 *
 * @param {Object} blockSelected - The selected block for which children are to be retrieved.
 * @param {Object} blocks - The blocks object containing various block types.
 * @returns {Array} An array of child blocks cloned from the blocks object.
 */
export const getChildrenFromList = (blockSelected, blocks) => {
  if (blockSelected.children) {
    return blockSelected.children.map((key) => cloneDeep(blocks[key]));
  }
  return [];
};

/**
 * Filters blocks based on the provided searchText.
 *
 * @param {string} searchText - The text to filter blocks.
 * @param {Object} blocks - The object containing blocks.
 * @returns {Object} - Filtered blocks.
 */
export const filterBlocksByText = (searchText, blocks) => {
  if (!searchText) {
    return {};
  }
  const copyBlocks = cloneDeep(blocks);
  return Object.keys(copyBlocks).reduce((result, key) => {
    const item = copyBlocks[key];
    if (item.path.toLowerCase().includes(searchText.toLowerCase())) {
      result[key] = item;
    }
    return result;
  }, {});
};

/**
 * Formats a block path into title and subtitle.
 *
 * @param {string} path - The path to be formatted.
 * @returns {Object} - Formatted block path with title and subtitle.
 */
export const formatBlockPath = (path) => {
  if (!path) {
    return {
      title: '',
      subTitle: '',
    };
  }
  const pathSlitted = path.split(' / ');
  let title = pathSlitted.pop();
  const subTitle = pathSlitted.join(' / ');

  if (!title.trim()) {
    // If the last part is empty or contains only whitespace
    title = pathSlitted.pop();
  }
  return {
    title,
    subTitle,
  };
};

/**
 * Validates a URL using a regular expression.
 *
 * @param {string} url - The URL to be validated.
 * @returns {boolean} - True if the URL is valid, false otherwise.
 */
export const isValidURL = (url) => {
  // Regular expression for a basic URL validation
  const urlPattern = /^(https?|ftp):\/\/[^\s/$.?#].[^\s]*$/;

  // Test the provided URL against the pattern
  return urlPattern.test(url);
};
