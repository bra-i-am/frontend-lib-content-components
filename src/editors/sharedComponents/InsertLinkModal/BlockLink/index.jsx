import PropTypes from 'prop-types';
import { Button } from '@edx/paragon';
import { LinkOff } from '@edx/paragon/icons';
import { formatBlockPath } from '../utils';

import './index.scss';

const BlockLink = ({ path, onCloseLink }) => {
  const { title, subTitle } = formatBlockPath(path);
  return (
    <div className="link-container d-flex row p-4 rounded border border-gray-400 mx-4 mt-3">
      <div className="col-10">
        <p className="text-left">{subTitle}</p>
        <p className="h2 w-20 title">{title}</p>
      </div>
      <div className="col-2">
        <Button
          variant="tertiary"
          className="d-flex justify-content-center align-items-center"
          data-testid="close-link-button"
          size="lg"
          iconBefore={LinkOff}
          onClick={onCloseLink}
        >
         &nbsp;
        </Button>
      </div>
    </div>
  );
};

BlockLink.defaultProps = {
  onCloseLink: () => {},
};

BlockLink.propTypes = {
  path: PropTypes.string.isRequired,
  onCloseLink: PropTypes.func,
};

export default BlockLink;
