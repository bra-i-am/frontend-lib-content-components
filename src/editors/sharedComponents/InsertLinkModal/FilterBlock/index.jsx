import { Button } from '@edx/paragon';
import PropTypes from 'prop-types';

import { formatBlockPath } from '../utils';

const FilterBlock = ({ block, onBlockFilterClick }) => {
  const { title, subTitle } = formatBlockPath(block.path);

  const handleBlockClick = () => {
    onBlockFilterClick(block);
  };

  return (
    <Button
      key={`filtered_block_${block.id}`}
      variant="tertiary"
      className="d-flex flex-column w-100 align-items-start p-3"
      onClick={handleBlockClick}
      data-testid="filter-block-item"
    >
      <span className="h5 text-left">{subTitle}</span>
      <span className="h3">{title}</span>
    </Button>
  );
};

FilterBlock.defaultProps = {
  onBlockFilterClick: () => {},
};

const blockShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  blockId: PropTypes.string.isRequired,
  lmsWebUrl: PropTypes.string.isRequired,
  legacyWebUrl: PropTypes.string.isRequired,
  studentViewUrl: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  displayName: PropTypes.string.isRequired,
  children: PropTypes.arrayOf(PropTypes.string),
});

FilterBlock.propTypes = {
  block: PropTypes.objectOf(blockShape).isRequired,
  onBlockFilterClick: PropTypes.func,
};

export default FilterBlock;
