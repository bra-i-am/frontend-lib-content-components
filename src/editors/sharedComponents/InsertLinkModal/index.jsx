import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { logError } from '@edx/frontend-platform/logging';
import { useIntl } from '@edx/frontend-platform/i18n';
import {
  Button,
  Tabs,
  Tab,
  Form,
} from '@edx/paragon';
import BaseModal from '../BaseModal';
import BlocksList from './BlocksList';
import BlockLink from './BlockLink';
import SearchBlocks from './SearchBlocks';
import { formatBlocks, isValidURL } from './utils';
import { getBlocksFromCourse } from './api';

import messages from './messages';
import './index.scss';

const InsertLinkModal = ({
  courseId,
  isOpen,
  onClose,
  editorRef,
}) => {
  const intl = useIntl();
  const [searchField, setSearchField] = useState('');
  const [blocksSearched, setBlocksSearched] = useState(false);
  const [blockSelected, setBlocksSelected] = useState(null);
  const [blocksList, setBlocksList] = useState(null);
  const [invalidUrlInput, setInvalidUrlInput] = useState(false);
  const [inputUrlValue, setInputUrlValue] = useState('');

  const handleSearchedBlocks = (isSearched) => {
    setBlocksSearched(isSearched);
  };

  const handleSelectedBlock = (blockSelectedFromList) => {
    setBlocksSelected(blockSelectedFromList);
    setInputUrlValue('');
  };

  const handleCloseLink = () => {
    setSearchField('');
    setBlocksSelected(null);
  };

  const handleChangeInputUrl = ({ target: { value } }) => {
    setInputUrlValue(value);
  };

  /* istanbul ignore next */
  const handleSave = () => {
    const editor = editorRef.current;
    const urlPath = blockSelected?.lmsWebUrl || inputUrlValue;
    if (editor && urlPath) {
      const validateUrl = isValidURL(urlPath);

      if (!validateUrl) {
        setInvalidUrlInput(true);
        return;
      }

      const selectedText = editor.selection.getContent({ format: 'text' });

      if (selectedText.trim() !== '') {
        const linkHtml = `<a href="${urlPath}" data-mce-href="${urlPath}" target="_blank">${selectedText}</a>`;
        editor.selection.setContent(linkHtml);
      }
    }

    onClose();
  };

  useEffect(() => {
    const getBlocksList = async () => {
      try {
        const blocksData = await getBlocksFromCourse(courseId);
        const { blocks: blocksResponse, root: rootBlocksResponse } = blocksData;
        const blockListFormatted = formatBlocks(
          blocksResponse,
          rootBlocksResponse,
        );
        setBlocksList(blockListFormatted);
      } catch (error) {
        logError(error);
      }
    };

    getBlocksList();
  }, []);

  return (
    <BaseModal
      isOpen={isOpen}
      close={onClose}
      title={intl.formatMessage(messages.insertLinkModalTitle)}
      confirmAction={(
        <Button variant="primary" onClick={handleSave}>
          {intl.formatMessage(messages.insertLinkModalButtonSave)}
        </Button>
      )}
    >
      {blockSelected ? (
        <BlockLink path={blockSelected.path} onCloseLink={handleCloseLink} />
      ) : (
        <Tabs
          variant="tabs"
          defaultActiveKey="course-pages"
          id="uncontrolled-tab-example"
          className="mt-3 justify-content-around w-100"
        >
          <Tab
            eventKey="course-pages"
            title={intl.formatMessage(messages.insertLinkModalCoursePagesTabTitle)}
            className="col-12 w-100 tabs-container"
          >
            <SearchBlocks
              blocks={blocksList || {}}
              onSearchFilter={handleSearchedBlocks}
              searchInputValue={searchField}
              onBlockSelected={handleSelectedBlock}
            />
            {!blocksSearched && (
              <BlocksList
                blocks={blocksList || {}}
                onBlockSelected={handleSelectedBlock}
              />
            )}
          </Tab>
          <Tab
            eventKey="url"
            title={intl.formatMessage(messages.insertLinkModalUrlTabTitle)}
            className="col-12 tabs-container"
          >
            <Form.Group isInvalid={invalidUrlInput} className="my-4">
              <Form.Control
                placeholder={intl.formatMessage(messages.insertLinkModalInputPlaceholder)}
                onChange={handleChangeInputUrl}
              />
              {invalidUrlInput && (
                <Form.Control.Feedback type="invalid">
                  {intl.formatMessage(messages.insertLinkModalInputErrorMessage)}
                </Form.Control.Feedback>
              )}
            </Form.Group>
          </Tab>
        </Tabs>
      )}
    </BaseModal>
  );
};

InsertLinkModal.propTypes = {
  courseId: PropTypes.string.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  editorRef: PropTypes.shape({
    current: PropTypes.shape({
      selection: PropTypes.shape({
        getContent: PropTypes.func,
        setContent: PropTypes.func,
      }),
    }),
  }).isRequired,
};

export default InsertLinkModal;
