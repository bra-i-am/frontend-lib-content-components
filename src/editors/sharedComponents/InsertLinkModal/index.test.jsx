import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render, screen } from '@testing-library/react';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import { logError } from '@edx/frontend-platform/logging';

import * as api from './api';

import InsertLinkModal from '.';

jest.mock('@edx/frontend-platform/logging', () => ({
  logError: jest.fn(),
}));

jest.mock('./api', () => ({
  getBlocksFromCourse: jest.fn().mockResolvedValue({
    blocks: {},
    root: {},
  }),
}));

jest.mock('./utils', () => ({
  ...jest.requireActual('./utils'),
  formatBlocks: jest.fn(),
  isValidURL: jest.fn(),
}));

jest.unmock('@edx/frontend-platform/i18n');
jest.unmock('@openedx/paragon');
jest.unmock('@openedx/paragon/icons');

describe('InsertLinkModal', () => {
  const mockProps = {
    courseId: 'course123',
    isOpen: true,
    onClose: jest.fn(),
    editorRef: {
      current: {
        selection: {
          getContent: jest.fn(),
          setContent: jest.fn(),
        },
      },
    },
  };

  // eslint-disable-next-line react/prop-types
  const IntlProviderWrapper = ({ children }) => (
    <IntlProvider locale="en">{children}</IntlProvider>
  );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = (overrideProps = {}) => render(
    <IntlProviderWrapper>
      <InsertLinkModal {...mockProps} {...overrideProps} />
    </IntlProviderWrapper>,
  );

  test('snapshot', async () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('renders without crashing', async () => {
    renderComponent();
    expect(screen.getByText('Link to')).toBeInTheDocument();
  });

  test('should show Course pages and URL tabs', () => {
    renderComponent();

    const tabs = screen.getAllByRole('tab');
    const [coursePagesTab, urlTab] = tabs;

    expect(coursePagesTab).toHaveTextContent('Course pages');
    expect(urlTab).toHaveTextContent('URL');
  });

  test('should find Cancel and Save buttons', () => {
    renderComponent();

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();

    const saveButton = screen.getByText('Save');
    expect(saveButton).toBeInTheDocument();
  });

  test('should show input for url when URL tab is clicked', () => {
    const { getByTestId } = renderComponent();

    const tabs = screen.getAllByRole('tab');
    const [, urlTab] = tabs;

    fireEvent.click(urlTab);

    const urlInput = getByTestId('url-input');
    expect(urlInput).toBeInTheDocument();
  });

  test('should show message when the url is invalid', () => {
    const { getByTestId, getByText } = renderComponent();

    const tabs = screen.getAllByRole('tab');
    const [, urlTab] = tabs;

    fireEvent.click(urlTab);

    const urlInput = getByTestId('url-input');
    fireEvent.change(urlInput, { target: { value: 'invalid-url' } });

    const saveButton = getByText('Save');
    fireEvent.click(saveButton);

    const errorMessage = getByText('The url provided is invalid');
    expect(errorMessage).toBeInTheDocument();
  });

  test('should call logError on API error', async () => {
    api.getBlocksFromCourse.mockRejectedValue(new Error('API error'));

    renderComponent();

    // Wait for the useEffect to complete
    await new Promise((resolve) => setTimeout(resolve, 0));

    expect(logError).toHaveBeenCalledWith(new Error('API error'));
  });
});
