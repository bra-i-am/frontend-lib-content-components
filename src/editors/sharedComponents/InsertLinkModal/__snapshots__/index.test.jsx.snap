// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsertLinkModal snapshot 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  fallbackOnEmptyString={true}
  formats={Object {}}
  locale="en"
  messages={Object {}}
  onError={[Function]}
  onWarn={[Function]}
  textComponent={Symbol(react.fragment)}
>
  <InsertLinkModal
    courseId="course123"
    editorRef={
      Object {
        "current": Object {
          "selection": Object {
            "getContent": [MockFunction],
            "setContent": [MockFunction],
          },
        },
      }
    }
    isOpen={true}
    onClose={[MockFunction]}
  />
</IntlProvider>
`;
